pageserver/src/tenant.rs

```diff
@@ -1639,64 +1644,88 @@ impl Tenant {
+            let mut activated_timelines = 0;
 
-                    for timeline in not_broken_timelines {
-                        timeline.activate(broker_client.clone(), ctx);
-                        activated_timelines += 1;
-                    }
+            for timeline in not_broken_timelines {
+                timeline.activate(broker_client.clone(), ctx);
+                activated_timelines += 1;
+            }
 
-                    let elapsed = self.loading_started_at.elapsed();
-                    let total_timelines = timelines_accessor.len();
+            self.state.send_modify(move |current_state| {
+                assert!(
+                    *current_state == TenantState::Activating,
+                    "set_stopping and set_broken wait for us to leave Activating state",
+                );
+                *current_state = TenantState::Active;
 
-                    // log a lot of stuff, because some tenants sometimes suffer from user-visible
-                    // times to activate. see https://github.com/neondatabase/neon/issues/4025
-                    info!(
-                        since_creation_millis = elapsed.as_millis(),
-                        tenant_id = %self.tenant_id,
-                        activated_timelines,
-                        total_timelines,
-                        post_state = <&'static str>::from(&*current_state),
-                        "activation attempt finished"
-                    );
-                }
-            }
-        });
-        result
+                let elapsed = self.loading_started_at.elapsed();
+                let total_timelines = timelines_accessor.len();
+
+                // log a lot of stuff, because some tenants sometimes suffer from user-visible
+                // times to activate. see https://github.com/neondatabase/neon/issues/4025
+                info!(
+                    since_creation_millis = elapsed.as_millis(),
+                    tenant_id = %self.tenant_id,
+                    activated_timelines,
+                    total_timelines,
+                    post_state = <&'static str>::from(&*current_state),
+                    "activation attempt finished"
+                );
+            });
+        };
+        Ok(())
     }
 
-    /// Change tenant status to Stopping, to mark that it is being shut down
-    pub fn set_stopping(&self) {
+    /// Change tenant status to Stopping, to mark that it is being shut down.
+    ///
+    /// This function is not cancel-safe!
+    pub async fn set_stopping(&self) {
+        // Get the rx before checking state inside send_if_modified.
+        // This way, when we later rx.changed().await, we won't have missed
+        // any state changes.
+        let mut rx = self.state.subscribe();
+        while *rx.borrow() == TenantState::Activating {
+            rx.changed()
+                .await
+                .expect("we're a method on Tenant, so, we're keeping self.state alive here");
+        }
```

Comment:
This is exactly why I dislike using watch channels.

From the `Receiver::borrow(&self)` docs:

> This method does not mark the returned value as seen, so future calls to [changed](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed) may return immediately even if you have already seen the value with a call to borrow.

I think we should instead:

```suggestion
        while *rx.borrow_and_update() == TenantState::Activating {
            rx.changed()
                .await
                .expect("we're a method on Tenant, so, we're keeping self.state alive here");
        }
```

With `Receiver::borrow_and_update`:

> This method marks the current value as seen. Subsequent calls to [changed](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.changed) will not return immediately until the [Sender](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html) has modified the shared value again.

Related part about `Receiver::changed`:

> If the newest value in the channel has not yet been marked seen when this method is called, the method marks that value seen and returns immediately. If the newest value has already been marked seen, then the method sleeps until a new message is sent by the [Sender](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html) connected to this Receiver, or until the [Sender](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html) is dropped.

So maybe it's just a single extra iteration? Well, my confusion here is why I dislike watch with borrow + changed, but it with 1.28.1 now has better api available:

```suggestion
        {
            rx.wait_for(|state| state != TenantState::Activating)
                .await
                .expect("cannot drop self.state while on a &self method");
        }
```

Extra block for shortening the lifetime of the returned Ref type. I don't know how long it could get extended, but this way it surely will not be.

Reply:
Nice, I didn't know there's `wait_for` , will switch

Reply:
fe4ef121b6c4ffeaf9175001c463cfdc598164d1

Reply:
Oki understanding the original issue for `Receiver::wait_for` through the feature impl tokio PR 5611 (intentionally not linking), this usage of `borrow + changed` was correct after all. Would still prefer the common `self.wait_past_activating().await;` or something similar, where we could document the:

```
(attaching|loading)
  (activating
    (active
      (stopping (broken)?|broken)
    |broken)
  |broken)
```

I think the regex matches the drawnings, but it doesn't look good.

Reply:
Resolving, mermaid is not a reason to hold this PR back.